ocaml/xapi/test_guest_agent.ml

```diff
@@ -0,0 +1,153 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open OUnit
+open Test_highlevel
+
+module Networks = Generic.Make (struct
+    module Io = struct
+      type input_t = string list
+      type output_t = (string * string) list
+
+      let string_of_input_t = Test_printers.(list string)
+      let string_of_output_t = Test_printers.(assoc_list string string)
+    end
+
+    type 'a tree = T of 'a * 'a tree list
+
+    let rec add_path_to_tree (T(root, children)) = function
+      | [] -> (T(root, children))
+      | node :: rest_of_path ->
+        try 
+          let T(_, children_of_node) = List.find (fun (T(n, _)) -> n = node) children in
+          let T(node, children_of_node) = add_path_to_tree (T(node, children_of_node)) rest_of_path in
+          T(root, (T(node, children_of_node)) :: (List.filter (fun (T(n, _)) -> n <> node) children))
+        with Not_found ->
+          T(root, (add_path_to_tree (T(node, [])) rest_of_path) :: children)
+
+    let construct_tree tree path =
+      let open Stdext.Xstringext in
+      let nodes = String.split_f (fun s -> s = '/') path in
+      add_path_to_tree tree nodes
+
+    let rec list_helper children = function
+      | [] -> List.map (fun (T(node, _)) -> node) children
+      | node :: rest_of_path ->
+        try
+          let T(_, children_of_node) = List.find (fun (T(n, _)) -> n = node) children in
+          list_helper children_of_node rest_of_path
+        with Not_found -> []
+
+    let list (T(root, children)) path =
+      let open Stdext.Xstringext in
+      let nodes = String.split_f (fun s -> s = '/') path in
+      list_helper children nodes
+
+    
+    let transform input = 
+      let tree = List.fold_left construct_tree (T("", [])) input in
+      Xapi_guest_agent.networks "attr" (list tree)
+
+    let tests = [
+      (* basic cases *)
+      [ "attr/vif/0/ipv6/0";
+      ], [ "attr/vif/0/ipv6/0", "0/ipv6/0";
+      ];
+
+      [ "attr/vif/0/ipv4/0";
+      ], [ "attr/vif/0/ipv4/0", "0/ip";
+           "attr/vif/0/ipv4/0", "0/ipv4/0";
+      ];
+
+      [ "attr/eth0/ip";
+      ], [ "attr/eth0/ip", "0/ip";
+           "attr/eth0/ip", "0/ipv4/0";
```

Comment:
Thanks @robhoes 
Does this aim to make the test more clear?
I'm asking because there is no lookup call in the test target function Xapi_guest_agent.networks.
The value like "192.168.0.1" is looked up outside the target function.